system/mp/mpicomm.cpp

```diff
@@ -0,0 +1,362 @@
+#include <limits.h>
+
+#include <unistd.h>
+
+#include "mpicomm.hpp"
+#include "mpbuff.hpp"
+#include "mputil.hpp"
+#include "mplog.hpp"
+
+#include "mpi_wrapper.hpp"
+
+class NodeCommunicator: public ICommunicator, public CInterface
+{
+private:
+    class SelfMessage
+    {
+        mptag_t mptag;
+        CMessageBuffer* message = NULL;
+    public:
+        SelfMessage(mptag_t _tag, CMessageBuffer* _message): mptag(_tag), message(_message){};
+        bool isTag(mptag_t _tag)
+        {
+            return mptag == _tag;
+        }
+        CMessageBuffer *getMessage()
+        {
+            return message;
+        }
+        mptag_t tag()
+        {
+            return mptag;
+        }
+
+        ~SelfMessage()
+        {
+            if (message){
+                delete message;
+            }
+        }
+    };
+    Owned<IGroup> group;
+    rank_t myrank;
+    rank_t commSize;
+    MPI::Comm& comm;
+    std::vector<SelfMessage*> selfMessages;
+    CriticalSection selfMessagesLock;
+private:
+    void addSelfMessage(SelfMessage *selfMessage)
+    {
+        _TF("addSelfMessage", selfMessage->tag());
+        CriticalBlock block(selfMessagesLock);
+        selfMessages.push_back(selfMessage);
+    }
+
+    SelfMessage *popSelfMessage(mptag_t tag)
+    {
+        _TF("popSelfMessage", tag);
+        CriticalBlock block(selfMessagesLock);
+        int index = -1;
+        for(int i=0; i< selfMessages.size(); i++)
+        {
+            if (selfMessages[i]->isTag(tag))
+            {
+                index = i;
+                break;
+            }
+        }
+        SelfMessage *ret = NULL;
+        if (index != -1)
+        {
+            ret = selfMessages[index];
+            selfMessages.erase(selfMessages.begin()+index);
+        }
+        return ret;
+    }
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    bool send(CMessageBuffer &mbuf, rank_t dstrank, mptag_t tag, unsigned timeout)
+    {
+        _TF("send", dstrank, tag, mbuf.getReplyTag(), timeout);
+        assertex(dstrank!=RANK_NULL);
+        CTimeMon tm(timeout);
+        rank_t startrank = dstrank;
+        rank_t endrank;
+        if (dstrank==RANK_ALL || dstrank==RANK_ALL_OTHER)
+        {
+            startrank = 0;
+            endrank = commSize-1;
+        } else if (dstrank==RANK_RANDOM)
+        {
+            if (commSize>1)
+            {
+                do
+                {
+                    startrank = getRandom()%commSize;
+                } while (startrank==myrank);
+            } else
+            {
+                assertex(myrank!=0);
+                startrank = 0;
+            }
+            endrank = startrank;
+        } else {
+            endrank = startrank;
+        }
+        _T("commSize="<<commSize);
+        for (;startrank<=endrank;startrank++)
+        {
+            if (startrank==myrank)
+            {
+                if (dstrank !=RANK_ALL_OTHER)
+                {
+                    CMessageBuffer *ret = mbuf.clone();
```

Comment:
it's a pity to allocate a new object. It would be better if SelfMessage cloned content to it's copy.
Will need a new method in CMessageBuffer for that though.